src/server/channel_store.h

```diff
@@ -34,39 +46,108 @@ class ChannelStore {
+  // Initialize ChannelStore bound to specific control block.
+  ChannelStore(ControlBlock* cb);
 
-  std::vector<Subscriber> FetchSubscribers(std::string_view channel);
+  // Fetch all subscribers for channel, including matching patterns.
+  std::vector<Subscriber> FetchSubscribers(std::string_view channel) const;
 
   std::vector<std::string> ListChannels(const std::string_view pattern) const;
   size_t PatternCount() const;
 
  private:
   using ThreadId = unsigned;
+
+  // Subscribers for a single channel/pattern.
   using SubscribeMap = absl::flat_hash_map<ConnectionContext*, ThreadId>;
 
-  struct ChannelMap : absl::flat_hash_map<std::string, std::unique_ptr<SubscribeMap>> {
+  // Wrapper around atomic pointer that allows copying and moving.
+  // Made to overcome restrictions of absl::flat_hash_map.
+  // Copy/Move don't need to be atomic with RCU.
+  struct UpdatablePointer {
+    UpdatablePointer(SubscribeMap* sm) : ptr{sm} {
+    }
+
+    UpdatablePointer(const UpdatablePointer& other);
+
+    SubscribeMap* Get() const;
+    void Set(SubscribeMap* sm);
+
+    SubscribeMap* operator->();
+    const SubscribeMap& operator*() const;
+
+   private:
+    std::atomic<SubscribeMap*> ptr;
+  };
+
+  // SubscriberMaps for channels/patterns.
+  struct ChannelMap : absl::flat_hash_map<std::string, UpdatablePointer> {
     void Add(std::string_view key, ConnectionContext* me, uint32_t thread_id);
     void Remove(std::string_view key, ConnectionContext* me);
+
+    // Delete all stored SubscribeMap pointers.
+    void DeleteAll();
   };
 
+ private:
+  ChannelStore(ChannelMap* channels, ChannelMap* patterns, ControlBlock*);
+
   static void Fill(const SubscribeMap& src, const std::string& pattern,
                    std::vector<Subscriber>* out);
 
-  mutable folly::RWSpinLock lock_;
-  ChannelMap channels_;
-  ChannelMap patterns_;
+  ChannelMap* channels_;
+  ChannelMap* patterns_;
+  ControlBlock* control_block_;
+};
+
+// Performs RCU (read-copy-update) updates to the channel store.
+// See ChannelStore header top for design details.
+// Queues operations and performs them with Apply().
+struct ChannelStoreUpdater {
+  ChannelStoreUpdater(ChannelStore* store, bool pattern, ConnectionContext* cntx,
+                      uint32_t thread_id);
+
+  void Add(std::string_view key);
+  void Remove(std::string_view key);
+  void Apply();
+
+ private:
+  using ChannelMap = ChannelStore::ChannelMap;
+
+  // Get target map and flag whether it was copied.
+  std::pair<ChannelMap*, bool> GetTargetMap();
+
+  // Apply modifly operation to target map.
+  void Modify(ChannelMap* target, std::string_view key, bool add);
+
+ private:
+  ChannelStore* store_;
+
+  bool pattern_;
+  ConnectionContext* cntx_;
+  uint32_t thread_id_;
+
+  // Pending operations.
+  std::vector<std::pair<std::string_view, bool>> ops_;
+
+  // Replaced SubscribeMaps that need to be deleted safely.
+  std::vector<ChannelStore::SubscribeMap*> freelist_;
 };
```

Comment:
I just noticed the bool to_add (and I didn't comment what it means) is always the same for all ops